rest/oidc_api_test.go

```diff
@@ -1,139 +1,865 @@
+			name:    "oidc default provider specified and current provider is default",
+			authURL: "/db/_oidc?provider=foo&offline=true",
+			issuer:  issuerFoo,
+		}, {
+			name:    "oidc challenge default provider specified and current provider is default",
+			authURL: "/db/_oidc_challenge?provider=foo&offline=true",
+			issuer:  issuerFoo,
+		}, {
+			// When multiple providers are defined, default provider is specified and no current provider is
+			// provided, then provider name should NOT be added to the generated OpenID Connect callback URL.
+			name:    "oidc default provider specified with no current provider",
+			authURL: "/db/_oidc?offline=true",
+			issuer:  issuerFoo,
+		}, {
+			name:    "oidc challenge default provider specified with no current provider",
+			authURL: "/db/_oidc_challenge?offline=true",
+			issuer:  issuerFoo,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			mockAuthServer.SetOption("issuer", tc.issuer)
+			var location string
+			if strings.Contains(tc.authURL, "_oidc_challenge") {
+				resp := rt.SendAdminRequest(http.MethodGet, tc.authURL, "")
+				require.Equal(t, http.StatusUnauthorized, resp.Code)
+				wwwAuthHeader := resp.Header().Get("Www-Authenticate")
+				location = regexp.MustCompile(`login="(?P<login>.*?)"`).FindStringSubmatch(wwwAuthHeader)[1]
+			} else {
+				resp := rt.SendAdminRequest(http.MethodGet, tc.authURL, "")
+				require.Equal(t, http.StatusFound, resp.Code)
+				location = resp.Header().Get(headerLocation)
+			}
+			require.NotEmpty(t, location, "error extracting location from header")
+			locationURL, err := url.Parse(location)
+			require.NoError(t, err, "error parsing location URL")
+			redirectURI := locationURL.Query().Get(requestParamRedirectURI)
+			redirectURL, err := url.Parse(redirectURI)
+			require.NoError(t, err, "error parsing redirect_uri URL")
+			assert.Equal(t, tc.wantProvider, redirectURL.Query().Get(auth.OIDCAuthProvider))
+		})
+	}
+}
+
+// Checks End to end OIDC Authorization Code flow.
+func TestOpenIDConnectAuth(t *testing.T) {
+	mockAuthServer := newMockAuthServer()
+	mockAuthServer.Start()
+	defer mockAuthServer.Shutdown()
+
+	issuerFoo := mockAuthServer.URL + "/foo"
+	mockAuthServer.SetOption("issuer", issuerFoo)
+
+	type test struct {
+		name                      string
+		providers                 auth.OIDCProviderMap
+		defaultProvider           string
+		authURL                   string
+		register                  bool
+		includeAccessToken        bool
+		notConfiguredProvider     bool
+		wantUsername              string
+		wantNoCode                bool
+		wantCallbackError         bool
+		wantUntoldProvider        bool
+		wantDiscoveryError        bool
+		wantCallbackTokenExgError bool
+		wantRefreshTokenExgError  bool
+		wantCallbackNoIDToken     bool
+		wantRefreshNoIDToken      bool
+	}
+	tests := []test{
+		{
+			name: "new user auto registration enabled single provider",
```

Comment:
The test 'name' property in these tests (and others) isn't concise enough for a name, but isn't verbose enough to be a useful description.  For example, on initial scan, it's not obvious whether this is 'auto registration enabled', or 'enabled single provider'.  What do you think about having a more concise name property, and more verbose description property?